### ISSUE TITLE
Support authenticated urls for http_archive (.netrc or other)

### ISSUE THREAD

**Issue by yeswalrus** [facebook/buck2#281]
I'm was attempting to migrate mid-size a Bazel repo over, but immediately hit a key sticking point - I need to be able to fetch pre-built binaries from an authenticated artifactory instance. Bazel simply allows using .netrc (similar to `curl -c`), which works quite well and is standard. Having similar support would be ideal, but anything allowing access to authenticated remote urls is a requirement for us.

**Comment by ndmitchell** [comment]
Can you just define a genrule that calls curl? You'll need to tag the rule as having to run locally if you have remote execution set up.

**Comment by yeswalrus** [comment]
As a workaround yes, but I do think it would be beneficial to have more built-in support for this.

**Comment by stepancheg** [comment]
> As a workaround yes, but I do think it would be beneficial to have more built-in support for this.

It is not trivial. Some people want to authenticate with passwords, other with certificates, third want support OAuth, so so on, and while committing to support of plethora of authentication methods would indeed be beneficial to users, maintaining would it would be not easy.

However, perhaps buck2 may support `curl` builtin rule in prelude, which would just invoke system curl. That rule can be parameterized with flags supported by `curl`.

**Comment by cjhopman** [comment]
Yeah, I guess I'm more sympathetic to the idea of supporting some standard ways of doing authentication for http_file/archive/etc. Certainly we wouldn't want to have a goal of supporting everything that people might possibly want, but supporting one or two of the most common and easy to use things seems pretty reasonable to me.

**Comment by yeswalrus** [comment]
Yeah, I'm really looking for something like point 2, where a user can have some credentials stored in a config file or env var (not in git obviously) and have those credentials matched with particular URL patterns. It would be super helpful to have this as a builtin. I'd love to try and convince my org to adopt Buck2, but a lack of features like this would be a big barrier to adoption

**Comment by ndmitchell** [comment]
What do other projects do in this space? I think Bazel just doesn't have a download_file. It seems there's a fairly big design space, and if we could just cheat by copying from someone, that would be easiest.
